### PR TITLE
[4.0] Checksum extensions - GSoC Expand Extensions Manager

### PR DESCRIPTION
**Pull request by NunoLopesPT**
###  Introduction

Currently there is no security when downloading files from update sites or uploading a package, with this feature the main goal is to make sure that only the original files are downloaded and installed by checking the integrity of the package(SHA256, SHA1 and MD5 Hashes), lowering the risk of getting infected files that can risk the user.

### Summary of Changes
This verification will only be made for installing packages from URL or uploading a package file in the Install view, this is how the process of verification will happen:
screenshot from 2017-08-08 16-53-23
(note: I forgot to place the SHA-256 hash here)

### Testing Instructions

We will have 3 packages to test:

**Package with the correct hashes in the update server manifest:**
[component_joomla.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1212365/component_joomla.zip)

Here is the update server manifest:
https://www.jah-tz.de/downloads/core/gsoc17/extension.xml

**Package with the wrong hashes:**
[component_joomla_wrong_checksum.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1212366/component_joomla_wrong_checksum.zip)

I remember here that without the Force Install checked you won't install the extension

Update server manifest:
https://www.jah-tz.de/downloads/core/gsoc17/extension_wrong_hash.xml

**Package without hashes:**
[component_joomla_no_checksum.zip](https://github.com/joomla-projects/gsoc17_expand_extension_manager/files/1212367/component_joomla_no_checksum.zip)

Update server manifest:
https://www.jah-tz.de/downloads/core/gsoc17/extension_no_chechsum.xml

### Expected result

Case 1 - File Checksum OK:
A success message when the checksums are equal

Case 2 - File Checksum Failed:
A danger message when the checksums are not equal and the user does not want to force the installation, redirecting back to the view without installing the extension

Case 3 - File Checksum Failed but user wants to force install:
There will be a checkbox on the upload package and install from URL tab where the user can check if he really wants to install the extension even if the Checksum fails.
A warning if the checksum fails will appear.

Case 4 - No checksum found:
If the extension has no update site or no checksums (MD5, SHA1 or SHA256 tags) are showed in the update site manifest a warning should appear to make sure the users know that no security verification was provided in the extension package.
A info message will appear that no Hashes are available.

### Actual result

Currently there is no security or information related to this

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 1]
```diff
@@ -63,6 +63,10 @@ COM_INSTALLER_HEADING_UPDATESITE_NAME_ASC="Update Site ascending"
 COM_INSTALLER_HEADING_UPDATESITE_NAME_DESC="Update Site descending"
 COM_INSTALLER_HEADING_UPDATESITEID="ID"
 COM_INSTALLER_INSTALL_BUTTON="Install"
+COM_INSTALLER_INSTALL_CHECKSUM_NOT_FOUND="There were no checksums provided in the package"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]: Please add a period at the end of the sentence

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 2]
```diff
@@ -63,6 +63,10 @@ COM_INSTALLER_HEADING_UPDATESITE_NAME_ASC="Update Site ascending"
...
+COM_INSTALLER_INSTALL_CHECKSUM_WRONG="File Checksum Failed"
+COM_INSTALLER_INSTALL_CHECKSUM_CORRECT="File Checksum OK"
+COM_INSTALLER_INSTALL_CHECKSUM_WRONG_NO_INSTALL="File Checksum Failed, Extension was not installed"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]: Please add a period at the end of the sentence

File: administrator/language/en-GB/en-GB.plg_installer_urlinstaller.ini [hunk 3]
```diff
@@ -4,6 +4,8 @@
 ; Note : All ini files need to be saved as UTF-8
 
 PLG_INSTALLER_URLINSTALLER_BUTTON="Check and Install"
+PLG_INSTALLER_URLINSTALLER_FORCE_INSTALL="Force Install"
+PLG_INSTALLER_URLINSTALLER_FORCE_INSTALL_DETAILS="Forces the installation of the package even if it is not trusted"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_installer_urlinstaller.ini]: Please add a period at the end of the sentence

**Comment by alikon** [comment]
a light port to 3.8 #17619

**Comment by bembelimen** [comment]
Perhaps I did not understand the patch, but what problem do you want to fix with this PR?

Is the checksum just some random hash?

Which scenario does now have a better protection with this PR?

**Comment by brianteeman** [comment]
The provided hash lets you double-check that the file you downloaded was not corrupted accidentally in transit, or that the file you downloaded from another source (a faster mirror or github etc) is the same as the file available for download at the original website where the hash is published

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 4]
```diff
@@ -63,6 +63,10 @@ COM_INSTALLER_HEADING_UPDATESITE_NAME_ASC="Update Site ascending"
 COM_INSTALLER_HEADING_UPDATESITE_NAME_DESC="Update Site descending"
 COM_INSTALLER_HEADING_UPDATESITEID="ID"
 COM_INSTALLER_INSTALL_BUTTON="Install"
+COM_INSTALLER_INSTALL_CHECKSUM_NOT_FOUND="There were no checksums provided in the package."
+COM_INSTALLER_INSTALL_CHECKSUM_WRONG="File Checksum Failed"
+COM_INSTALLER_INSTALL_CHECKSUM_CORRECT="File Checksum OK"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_installer.ini]: Sort alpha order.

File: plugins/installer/packageinstaller/tmpl/default.php [hunk 5]
```diff
@@ -19,6 +19,14 @@
 <hr>
 
 <div id="uploader-wrapper">
+	<div class="control-group">
+		<label for="force_install" class="control-label" title="<?php echo \JText::_('PLG_INSTALLER_URLINSTALLER_FORCE_INSTALL_DETAILS') ?>">
```
Comment (Quy) [inline plugins/installer/packageinstaller/tmpl/default.php]: Add semicolon after `echo \JText::_('PLG_INSTALLER_URLINSTALLER_FORCE_INSTALL_DETAILS')`

File: plugins/installer/urlinstaller/tmpl/default.php [hunk 6]
```diff
@@ -21,6 +21,14 @@
 		<input type="text" id="install_url" name="install_url" class="form-control input-xlarge" placeholder="https://">
 	</div>
 </div>
+<div class="control-group">
+	<label for="force_install" class="control-label" title="<?php echo \JText::_('PLG_INSTALLER_URLINSTALLER_FORCE_INSTALL_DETAILS') ?>">
```
Comment (Quy) [inline plugins/installer/urlinstaller/tmpl/default.php]: Add semicolon after `echo \JText::_('PLG_INSTALLER_URLINSTALLER_FORCE_INSTALL_DETAILS')`

File: administrator/components/com_installer/Helper/InstallerHelper.php [hunk 7]
```diff
@@ -157,4 +157,81 @@ public static function getStateOptions()
...
+	{
+		$app = \JFactory::getApplication();
+
+		$hashes = array("sha256", "sha384", "sha512");
+		$hashOnFile = false;
+
+		foreach ($hashes as $hash)
+		{
+			$hashPackage = hash_file($hash, $packagefile);
+			$hashRemote  = $updateServerManifest->$hash->_data;
```
Comment (wilsonge) [inline administrator/components/com_installer/Helper/InstallerHelper.php:209]: If there is a hash missing you get PHP Errors. To solve this above this line add

```
// If there is no hashes set for an extension continue
if (!property_exists($updateServerManifest, $hash))
{
	continue;
}
```

**Comment by wilsonge** [comment]
1. Can we fix conflicts please :)
2. The bad checksum extension still installs (we talked about this after the panel - all the checksums in that file aren't the ones you check for)

**Comment by schultz-it-solutions** [comment]
The Update Server manifests seem to no longer exist
https://www.jah-tz.de/downloads/core/gsoc17/extension.xml

Could you please direct me to an example (or description) of the changes in the manifest.

**Comment by sanderpotjer** [comment]
@roland-d while working on the extension manager features, can you please update this PR as well (or create a new one).

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 8]
```diff
@@ -63,6 +63,10 @@ COM_INSTALLER_HEADING_UPDATESITE_NAME_ASC="Update Site ascending"
 COM_INSTALLER_HEADING_UPDATESITE_NAME_DESC="Update Site descending"
 COM_INSTALLER_HEADING_UPDATESITEID="ID"
 COM_INSTALLER_INSTALL_BUTTON="Install"
+COM_INSTALLER_INSTALL_CHECKSUM_CORRECT="File Checksum OK"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_installer.ini:66]: String not used.

File: administrator/components/com_installer/Helper/InstallerHelper.php [hunk 9]
```diff
@@ -157,4 +157,81 @@ public static function getStateOptions()
...
+	 *
+	 * @return  boolean  if the installation can continue or not
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function isChecksumValid($packagefile, $updateServerManifest, $forceValidation)
+	{
+		$app = \JFactory::getApplication();
+
+		$hashes = array("sha256", "sha384", "sha512");
```
Comment (Quy) [inline administrator/components/com_installer/Helper/InstallerHelper.php:203]:
```suggestion
		$hashes = ['sha256', 'sha384', 'sha512'];
```

**Comment by uglyeoin** [comment]
Have @wilsonge comments been taken account of?  Is this ready to test yet?

**Comment by wilsonge** [comment]
No no use in testing this right now.

**Comment by roland-d** [comment]
This can be closed and picked up in another PR if someone is interested.

**Pull request closed without merging**